### PR TITLE
Fix support for pushing to registry in Python Image

### LINKED ISSUE #153: [Python] Docker Image build with registry fails
When building an image to push to a registry, I get this:
```
      File "/home/matthieu/.virtualenvs/whisbee-infra/lib/python3.7/site-packages/pulumi/output.py", line 89, in is_value_known
        return await is_known and not contains_unknowns(await future)
      File "/home/matthieu/.virtualenvs/whisbee-infra/lib/python3.7/site-packages/pulumi/output.py", line 194, in run
        return await cast(Awaitable[U], transformed)
      File "/home/matthieu/.virtualenvs/whisbee-infra/lib/python3.7/site-packages/pulumi_docker/docker.py", line 188, in do_build
        image_name, path_or_build_val, repository_url_val, log_resource, connect_to_registry, skip_push)
      File "/home/matthieu/.virtualenvs/whisbee-infra/lib/python3.7/site-packages/pulumi_docker/docker.py", line 267, in build_and_push_image_worker_async
        registry_output = pulumi.Output.from_input(connect_to_registry())
    TypeError: get_registry() missing 1 required positional argument: 'r'
    error: an unhandled error occurred: Program exited with non-zero exit code: 1
```
Indeed, in `image.py`:
```
            async def get_registry(r):
                return ImageRegistry(r.server, r.username, r.password)
```
there is an argument `r` here where there shouldn't be one, and we should use `registry`. A small botch on my part in the port, will send a PR.

### LINKED ISSUE #154: Python library uses from_input but that doesn't do what is expected
It appears that the Python `Image` implementation [uses `from_input`](https://github.com/pulumi/pulumi-docker/blob/4dd52eacce777190066b23a5f4041ab44e676d54/sdk/python/pulumi_docker/image.py#L164) in an attempt to deeply unwrap inputs inside of the args, and then [uses them directly](https://github.com/pulumi/pulumi-docker/blob/4dd52eacce777190066b23a5f4041ab44e676d54/sdk/python/pulumi_docker/image.py#L111). Unfortunately, this fails for me [at this line](https://github.com/pulumi/pulumi-docker/blob/4dd52eacce777190066b23a5f4041ab44e676d54/sdk/python/pulumi_docker/utils.py#L28), because the `base_image_name` is still an input and not a raw string (assigned [here](https://github.com/pulumi/pulumi-docker/blob/4dd52eacce777190066b23a5f4041ab44e676d54/sdk/python/pulumi_docker/image.py#L116)).

I'm not sure what the right fix is. Perhaps the args needs to be a map instead so it's deeply unwrapped? Or the function needs to use more targeted `apply` calls? @lukehoban, any suggestion?

This is the repro program, which attempts to build and publish to an ECR registry:

```python
import base64
import pulumi
import pulumi_aws as aws
import pulumi_docker as docker

# Create a private ECR registry.
repo = aws.ecr.Repository('my-repo')

# Get registry info (creds and endpoint) so we can build/publish to it.
def getRegistryInfo(rid):
    creds = aws.ecr.get_credentials(registry_id=rid)
    decoded = base64.b64decode(creds.authorization_token)
    parts = decoded.split(':')
    if len(parts) != 2:
        raise Exception("Invalid credentials")
    return docker.ImageRegistry(creds.proxy_endpoint, parts[0], parts[1])
registry = repo.registry_id.apply(getRegistryInfo)

# Build and publish the image.
image = docker.Image('my-image',
    build='app',
    image_name=repo.repository_url,
    registry=registry,
)

# Export the resulting base name in addition to the specific version pushed.
pulumi.export('baseImageName', image.base_image_name)
pulumi.export('imageName', image.image_name)
```

Related to all of this: https://github.com/pulumi/pulumi/pull/4178

### PR DESCRIPTION
**Pull request by lukehoban**
Also add test coverage.

Fixes #153.
Fixes #154.

### CODE REVIEW

**Comment by lukehoban** [comment]
@rienafairefr I didn't realize till I'd already gotten this fixed that you mentioned you were already looking into it in #153!  Sorry if I stepped on your toes here.  If you have a PR in progress, happy to take that and close this out.

**Comment by rienafairefr** [comment]
@lukehoban totally OK, I had started to look into it, but I think this incorporates all the bugs I had found. And your PR has examples & test so it's even better. +1!

File: sdk/python/pulumi_docker/image.py [hunk 1]
```diff
@@ -160,8 +165,9 @@ def check_tag(t: Optional[str]):
                 'registry_server': registry.server if registry else None
             }
 
-        image_args = ImageArgs(image_name, build, local_image_name, registry, skip_push)
-        image_data = pulumi.Output.from_input(image_args).apply(get_image_data)
+        image_data = pulumi.Output.all(image_name, build, local_image_name, registry, skip_push).apply(
+            lambda args: get_image_data(_ImageArgs(*args))
+        )
 
         self.image_name, self.registry_server = image_data['image_name'], image_data['registry_server']
```
Comment (rienafairefr) [inline sdk/python/pulumi_docker/image.py]: I get a typing error here, `self.registry_server` should be a `pulumi.Output[Optional[str]]` I think (that's what I did in my fork [here](https://github.com/rienafairefr/pulumi-docker/tree/fix-registry-push))

File: sdk/python/pulumi_docker/docker.py [hunk 2]
```diff
@@ -518,7 +516,7 @@ def login_to_registry(registry, log_resource: pulumi.Resource) -> Awaitable:
     else:
         log_ephemeral(f'Reusing existing login for {username}@{registry_name}', log_resource)
 
-    return login_result.login_command
```
Comment (rienafairefr) [inline sdk/python/pulumi_docker/docker.py]: Does this work ? With `login_result = any(filter(...` above `login_result` is a boolean.

**Pull request merged**